nav2_map_server/src/map_3d/map_saver_3d.cpp

```diff
@@ -0,0 +1,217 @@
+MapSaver<sensor_msgs::msg::PointCloud2>::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Activating");
+
+  // create bond connection
+  createBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver<sensor_msgs::msg::PointCloud2>::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // destroy bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver<sensor_msgs::msg::PointCloud2>::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  save_map_service_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+MapSaver<sensor_msgs::msg::PointCloud2>::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void MapSaver<sensor_msgs::msg::PointCloud2>::saveMapCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::SaveMap3D::Request> request,
+  std::shared_ptr<nav2_msgs::srv::SaveMap3D::Response> response)
+{
+  map_3d::SaveParameters save_parameters;
+  save_parameters.map_file_name = request->map_url;
+  save_parameters.as_binary = request->as_binary;
+  save_parameters.format = request->file_format;
+
+  response->result = saveMapTopicToFile(request->map_topic, request->origin_topic, save_parameters);
+}
+
+bool MapSaver<sensor_msgs::msg::PointCloud2>::saveMapTopicToFile(
+  const std::string & map_topic,
+  const std::string & origin_topic,
+  const map_3d::SaveParameters & save_parameters)
+{
+  // Local copies of map_topic and save_parameters that could be changed
+  std::string map_topic_loc = map_topic;
+  std::string origin_topic_loc = origin_topic;
+  map_3d::SaveParameters save_parameters_loc = save_parameters;
+
+  RCLCPP_INFO(
+    get_logger(), "Saving map from \'%s\' topic and origin from topic \'%s\' to \'%s\' file",
+    map_topic_loc.c_str(), origin_topic_loc.c_str(), save_parameters_loc.map_file_name.c_str());
+
+  try {
+    // Pointer to map message received in the subscription callback
+    sensor_msgs::msg::PointCloud2::SharedPtr pcd_map_msg = nullptr;
+
+    // Pointer to the origin message received in the subscription callback
+    geometry_msgs::msg::Pose::SharedPtr origin_msg = nullptr;
+
+    // Mutex for handling map_msg shared resource
+    std::recursive_mutex access;
+
+    // Correct map_topic_loc if necessary
+    if (map_topic_loc.empty()) {
+      map_topic_loc = "map";
+      RCLCPP_WARN(
+        get_logger(), "Map topic unspecified. Map messages will be read from \'%s\' topic",
+        map_topic_loc.c_str());
+    }
+
+    // Correct origin_topic_loc if necessary
+    if (origin_topic_loc.empty()) {
+      origin_topic_loc = map_topic_loc + "_origin";
```

Comment:
From the above comments - I think this should be removed or a new 3D map message should be made so that the PC and the origin are combined with the metadata required, this structure is not an acceptable map format / synchronous